File: 25.k-个一组翻转链表.cpp
/*
 * @lc app=leetcode.cn id=25 lang=cpp
 *
 * [25] K 个一组翻转链表
 */
#include "iostream"
#include "map"
#include "vector"

using namespace std;
// @lc code=start
// Definition for singly-linked list.

// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };

class Solution {
  public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        //
        if (!head || !head->next)
            return head;
        int n = k;
        ListNode *p = head;
        ListNode *temp = new ListNode();
        ListNode *ret = temp;
        ListNode *q = head;
        while (p) {
            while (--n && p) {
                p = p->next;
            }
            if (n != 0 || !p) {
                return ret->next;
            }
            q = p->next;
            p->next = nullptr;
            temp->next = reverseList(head);
            temp = head;
            head->next = q;
            head = head->next;
            p = q;
            n = k;
        }
        return ret->next;
    }

    ListNode *reverseList(ListNode *head) {
        if (!head || !head->next)
            return head;
        ListNode *cur = head;
        ListNode *next = head->next;
        cur->next = NULL;
        ListNode *p = cur->next, *q = next->next;
        while (next) {
            cur->next = p;
            q = next->next;
            next->next = cur;
            p = cur;
            cur = next;
            next = q;
        }
        return cur;
    }
};
// @lc code=end
// int main() {
//     Solution s;
//     ListNode l5(5);
//     ListNode l4(4, &l5);
//     ListNode l3(3, &l4);
//     ListNode l2(2, &l3);
//     ListNode l1(1, &l2);
//     s.reverseKGroup(&l1, 2);
// }

File: 25.k-个一组翻转链表.go
/*
 * @lc app=leetcode.cn id=25 lang=golang
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

 func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	var ret *ListNode
	var pre *ListNode

	for head != nil {
		i := k
		p := head
		for i > 1 && p != nil {
			p = p.Next
			i--
		}
		if p == nil {
			break
		}
		next := p.Next
		p.Next = nil
		tail := head
		head = reverse(head)
		if pre != nil {
			pre.Next = head
		}
		if ret == nil {
			ret = head
		}
		tail.Next = next
		pre = tail
		head = next
	}
	if ret == nil {
		ret = head
	}
	return ret
}

func reverse(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	var (
		pre  *ListNode = nil
		next *ListNode = nil
		cur  *ListNode = nil
	)
	pre = head
	cur = head.Next
	pre.Next = nil
	for cur != nil {
		next = cur.Next
		cur.Next = pre
		pre = cur
		cur = next
	}
	return pre
}
// @lc code=end

